### PR TITLE
Qwen3-Next Model Provider + Pretraining Support

### PR DESCRIPTION
**Pull request by adityavavreNVDA**
This PR adds Qwen3-Next model support and has a dependecy on Megatron-LM's dev branch. 
1. Pre-training loss curves match Megatron-LM's training curves as shown below:
image

2. Unfused Attention Pretraining logs comparison with MLM
image

3. Able to successfully store a checkpoint and resume training from where it left off. 

4. DeepEP works in 2509 container with DeepEP package installed.

5. Add unit tests for model provider. [Done]

6. IMPORTANT EDIT: Qwen3-Next Pre-training uses Global Auxiliary Loss for load balancing. Updated the recipe and compared with MLM again:
image

### CODE REVIEW

File: src/megatron/bridge/models/qwen/qwen_provider.py [hunk 1]
```diff
@@ -401,3 +401,53 @@ class Qwen3MoEModelProvider235B_A22B(Qwen3MoEModelProvider):
...
+    linear_attention_type: str = "gated_delta_net" # Gated Delta Net used in 75% of the model layers
+    linear_attention_freq: Union[int, List[int]] = 4  # 1 gated standard attention layer per 4 layers
+    linear_conv_kernel_dim: int = 4
+    linear_key_head_dim: int = 128
+    linear_value_head_dim: int = 128
+    linear_num_key_heads: int = 16
+    linear_num_value_heads: int = 32
+
+@dataclass
+class Qwen3NextMoEModelProvider80B_A3B(Qwen3MoEModelProvider):
```
Comment (ananthsub) [inline src/megatron/bridge/models/qwen/qwen_provider.py]:
```suggestion
class Qwen3NextMoEModelProvider80B_A3B(Qwen3NextMoEModelProvider):
```
should inherit from Qwen3NextMoEModelProvider

File: src/megatron/bridge/models/qwen/qwen_provider.py [hunk 2]
```diff
@@ -401,3 +401,53 @@ class Qwen3MoEModelProvider235B_A22B(Qwen3MoEModelProvider):
...
+# Qwen 3-Next MoE Model Provider (based on GPTProvider)
+# =============================================================================
+
+
+@dataclass
+class Qwen3NextMoEModelProvider(Qwen3MoEModelProvider):
+    """Base provider for Qwen 3 MoE Models."""
+
+    layernorm_zero_centered_gamma: bool = True # Zero-centered RMSNorm
+    kv_channels: Optional[int] = 256
```
Comment (ananthsub) [inline src/megatron/bridge/models/qwen/qwen_provider.py]:
```suggestion
    kv_channels: int | None = 256
```

File: src/megatron/bridge/models/qwen/qwen_provider.py [hunk 3]
```diff
@@ -401,3 +401,53 @@ class Qwen3MoEModelProvider235B_A22B(Qwen3MoEModelProvider):
...
+    moe_token_dispatcher_type: str = "flex"
+    moe_shared_expert_gate: bool = True
+    moe_permute_fusion: bool = True
+    moe_router_fusion: bool = True
+    moe_enable_deepep: bool = True
+    moe_router_dtype: str = "fp32"
+
+    # Linear Attention specific parameters
+    linear_attention_type: str = "gated_delta_net" # Gated Delta Net used in 75% of the model layers
+    linear_attention_freq: Union[int, List[int]] = 4  # 1 gated standard attention layer per 4 layers
```
Comment (ananthsub) [inline src/megatron/bridge/models/qwen/qwen_provider.py]:
```suggestion
    linear_attention_freq: int | list[int] = 4  # 1 gated standard attention layer per 4 layers
```

File: src/megatron/bridge/training/utils/weight_decay_utils.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+def get_no_weight_decay_cond(
+    no_weight_decay_cond_type: str, default_skip_embedding_weight_decay: bool
+    ) -> Callable[[str, torch.Tensor], bool]:
+    """Get the no weight decay condition function."""
+
+    # Default case: no_wd_decay_cond_type is None
+    no_weight_decay_cond_fn = None
+
+    if no_weight_decay_cond_type == 'qwen3_next':
```
Comment (yaoyu-33) [inline src/megatron/bridge/training/utils/weight_decay_utils.py]: please do not make conditions like this, will be harder to maintain in the future.
you can just rename this function to `qwen3_next_no_weight_decay_cond`, make it only specific for qwen3-next. in the future if there are other models, they can also make their own function, not add more conditions.

File: src/megatron/bridge/training/setup.py [hunk 5]
```diff
@@ -179,11 +180,16 @@ def setup(
     )
     cfg.model.timers = timers
     cfg.optimizer.timers = timers
+    no_weight_decay_cond = get_no_weight_decay_cond(
```
Comment (yaoyu-33) [inline src/megatron/bridge/training/setup.py:190]: I would move this part to qwen3 receipes, i hope this can be set per model not here.
we can 
1. update optimizer cfg to make it contains "no_weight_decay_cond"
2. in recipe set `optimizer.no_weight_decay_cond = qwen3_next_no_weight_decay_cond`
3. in line 129 `no_weight_decay_cond=cfg.optimizer.no_weight_decay_cond`.

lemme also discuss this in our daily.

File: src/megatron/bridge/training/utils/weight_decay_utils.py (hunk 4 above)
Comment (adityavavreNVDA) [reply src/megatron/bridge/training/utils/weight_decay_utils.py]: How do you propose we do this then? We need to pass the no_weight_decay_cond function to setup_optimizer() in setup.py.

File: src/megatron/bridge/training/setup.py (hunk 5 above)
Comment (adityavavreNVDA) [reply src/megatron/bridge/training/setup.py:190]: Let me add these changes and test again.

Comment (yaoyu-33) [reply src/megatron/bridge/training/setup.py:190]: defer refactor to later per discussions.

File: src/megatron/bridge/training/setup.py [hunk 6]
```diff
@@ -179,11 +180,16 @@ def setup(
...
+        cfg.scheduler.no_weight_decay_cond_type,
```
Comment (yaoyu-33) [inline src/megatron/bridge/training/setup.py:191]: this needs to be guarded using getattr, I dont think all scheduler has the `no_weight_decay_cond_type`

File: src/megatron/bridge/recipes/qwen/qwen3_next.py [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+    base_output_dir = dir if dir is not None else os.path.join(os.getcwd(), "nemo_experiments")
+    run_output_dir = os.path.join(base_output_dir, name)
+    checkpoint_dir = os.path.join(run_output_dir, "checkpoints")
+    tensorboard_dir = os.path.join(run_output_dir, "tb_logs")
+
+    blend, blend_per_split, split = get_blend_fields_from_data_paths(
+        data_paths, data_args_path, train_data_path, valid_data_path, test_data_path, per_split_data_args_path, mock
+    )
+
+    # bridge = AutoBridge.from_hf_pretrained(hf_path)
```
Comment (yaoyu-33) [inline src/megatron/bridge/recipes/qwen/qwen3_next.py]: remove if no longer needed

File: src/megatron/bridge/training/setup.py (hunk 6 above)
Comment (adityavavreNVDA) [reply src/megatron/bridge/training/setup.py:191]: According to the config this is None by default. The get_no_weight_decay_cond function checks if this value is None and handles the default case. Would we still need getattr?

**Comment by adityavavreNVDA** [comment]
/ok to test [aabbceb](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/1027/commits/aabbceb4d9e83504a989111ec8efd390c622849f)

**Comment by adityavavreNVDA** [comment]
/ok to test [1459fcb](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/1027/commits/1459fcb71befab1a21d40c2f3f7964b08ac75fba)

**Comment by adityavavreNVDA** [comment]
/ok to test [8bf18c2](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/1027/commits/8bf18c2589aa0660f6cfddcb0bc470959719d89c)
